core/node/scrub/stream_scrub_task.go

```diff
@@ -0,0 +1,210 @@
+		With("spaceId", task.spaceId)
+	_, view, err := tp.cache.GetStream(tp.ctx, task.channelId)
+	if err != nil {
+		log.Error(
+			"Unable to scrub stream; could not fetch stream view",
+			"error",
+			err,
+		)
+		return
+	}
+
+	joinableView, ok := view.(events.JoinableStreamView)
+	if !ok {
+		log.Error("Unable to scrub stream; could not cast view to JoinableStreamView")
+		return
+	}
+
+	members, err := joinableView.GetChannelMembers()
+	if err != nil {
+		log.Error("Failed to fetch stream members", "error", err)
+		return
+	}
+	for member := range (*members).Iter() {
+		isEntitled, err := tp.chainAuth.IsEntitled(
+			tp.ctx,
+			tp.config,
+			auth.NewChainAuthArgsForChannel(
+				task.spaceId,
+				task.channelId,
+				member,
+				auth.PermissionRead,
+			),
+		)
+		if err != nil {
+			log.Error("Scrubbing error: unable to evaluate user entitlement",
+				"user",
+				member,
+				"error",
+				err,
+			)
+			continue
+		}
+		// In the case that the user is not entitled, they must have lost their entitlement
+		// after joining the channel, so let's go ahead and boot them.
+		if !isEntitled {
+			userId, err := AddressFromUserId(member)
+			if err != nil {
+				log.Error("Error converting user id into address", "member", member, "error", err)
+				continue
+			}
+			userStreamId, err := UserStreamIdFromBytes(userId)
+			if err != nil {
+				log.Error(
+					"Error constructing user id stream from user address",
+					"userAddress",
+					userId,
+					"error",
+					err,
+				)
+			}
+			log.Info("Entitlement loss detected; adding LEAVE event for user",
+				"user",
+				member,
+				"userStreamId",
+				userStreamId,
+			)
+			err = tp.eventAdder.AddEventPayload(
+				tp.ctx,
+				userStreamId,
+				events.Make_UserPayload_Membership(
+					MembershipOp_SO_LEAVE,
+					task.channelId,
+					&member,
+					task.spaceId[:],
+				),
+			)
+			if err != nil {
+				log.Error(
+					"scrub error: unable to add channel leave event to user stream",
+					"userStreamId",
+					userStreamId,
+					"error",
+					err,
+				)
+			}
+		}
+	}
+}
+
+type streamScrubTask struct {
+	channelId     StreamId
+	spaceId       StreamId
+	taskProcessor *streamScrubTaskProcessorImpl
+}
+
+func (t *streamScrubTask) process() {
+	t.taskProcessor.processTask(t)
+}
+
+// TryScheduleScrub schedules a stream scrub task if:
+// - the stream is a channel stream,
+// - the stream has not been recently scrubbed, and
+// - there is no pending scrub for the given stream.
+// If the worker pool is full, the method will not block but will return an error.
+// This is so that we don't affect ability to post updates to channels with stale scrubs
+// whenever the node falls behind due to high scrubbing volume.
+func (tp *streamScrubTaskProcessorImpl) TryScheduleScrub(
+	ctx context.Context,
+	streamId StreamId,
+) (bool, error) {
+	log := dlog.FromCtx(ctx).With("Func", "TryScheduleScrub").With("streamId", streamId)
+	if !ValidChannelStreamId(&streamId) {
+		return false, nil
+	}
```

Comment:
Just a note - if we ever start scrubbing spaces we'll need to make sure we kick the user from all the channels in the space before kicking from the space.

Reply:
I added a note here in the method as a reminder.

Reply:
ty